recordbatch.go

```diff
@@ -0,0 +1,104 @@
+package kafka
+
+import (
+	"bytes"
+	"time"
+)
+
+const recordBatchHeaderSize int32 = 0 +
+	8 + // base offset
+	4 + // batch length
+	4 + // partition leader epoch
+	1 + // magic
+	4 + // crc
+	2 + // attributes
+	4 + // last offset delta
+	8 + // first timestamp
+	8 + // max timestamp
+	8 + // producer id
+	2 + // producer epoch
+	4 + // base sequence
+	4 // msg count
+
+func recordBatchSize(msgs ...Message) (size int32) {
+	size = recordBatchHeaderSize
+	baseTime := msgs[0].Time
+
+	for i := range msgs {
+		msg := &msgs[i]
+		msz := recordSize(msg, msg.Time.Sub(baseTime), int64(i))
+		size += int32(msz + varIntLen(int64(msz)))
+	}
+
+	return
+}
+
+func compressRecordBatch(codec CompressionCodec, msgs ...Message) (compressed *bytes.Buffer, attributes int16, size int32, err error) {
+	compressed = acquireBuffer()
+	compressor := codec.NewWriter(compressed)
+	wb := &writeBuffer{w: compressor}
+
+	for i, msg := range msgs {
+		wb.writeRecord(0, msgs[0].Time, int64(i), msg)
+	}
+
+	if err = compressor.Close(); err != nil {
+		releaseBuffer(compressed)
+		return
+	}
+
+	attributes = int16(codec.Code())
+	size = recordBatchHeaderSize + int32(compressed.Len())
+	return
+}
+
+type recordBatch struct {
+	// required input parameters
+	codec      CompressionCodec
+	attributes int16
+	msgs       []Message
+
+	// parameters calculated during init
+	compressed *bytes.Buffer
+	size       int32
+}
+
+func (r *recordBatch) init() (err error) {
```

Comment:
It looks like it's never valid to use a `recordBatch` without calling `init`, how about making it a constructor instead?
```go
func newRecordBatch(codec CompressionCodec, msgs []Message) *recordBatch {
    ...
}
```
One challenge is managing the error returned by `compressRecordBatch`, maybe the `recordBatch` type can hold the error value like we do for `writeBuffer`?

Reply:
Yeah, I was choosing between this option and the option that I implemented. My motivation was to not duplicate the fields in the record batch and in the declaration of the method. I don't have any strong attachment to either solution. Would you prefer your variant?

Reply:
Since a `recordBatch` value is never fully initialized without calling `init` I think it would make the code harder to use wrong to have the construction contained in a single function 👍

Reply:
Done